src/webgpu/api/operation/storage_texture/read_write.spec.ts

```diff
@@ -0,0 +1,405 @@
+export const description = `
+Tests for the behavior of read-write storage textures.
+
+TODO:
+- Test resource usage transitions with read-write storage textures
+- Test 1D and 3D textures
+`;
+
+import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { kTextureFormatInfo } from '../../../format_info.js';
+import { GPUTest } from '../../../gpu_test.js';
+import { align } from '../../../util/math.js';
+
+type ReadWriteStorageTextureFormat = 'r32float' | 'r32sint' | 'r32uint';
+type ShaderStageForReadWriteStorageTexture = 'fragment' | 'compute';
+
+const kReadWriteStorageTextureFormats = ['r32float', 'r32sint', 'r32uint'] as const;
+const kShaderStagesForReadWriteStorageTexture = ['fragment', 'compute'] as const;
+
+interface TestSuite {
+  GetExpectedData(
+    textureFormat: ReadWriteStorageTextureFormat,
+    width: number,
+    height: number,
+    arrayLayers: number,
+    initialData: ArrayBuffer
+  ): ArrayBuffer;
+
+  RecordCommandsToTransform(
+    device: GPUDevice,
+    commandEncoder: GPUCommandEncoder,
+    rwTexture: GPUTexture,
+    format: ReadWriteStorageTextureFormat
+  ): void;
+}
+
+// Test the use of read-write storage textures in compute pipelines. The functionality of
+// textureBarrier() is also tested in this test suite.
+class TestSuiteForCompute implements TestSuite {
+  GetExpectedData(
+    textureFormat: ReadWriteStorageTextureFormat,
+    width: number,
+    height: number,
+    arrayLayers: number,
+    initialData: ArrayBuffer
+  ): ArrayBuffer {
+    const bytesPerBlock = kTextureFormatInfo[textureFormat].bytesPerBlock;
+    const bytesPerRowAlignment = align(bytesPerBlock * width, 256);
+    const itemsPerRow = bytesPerRowAlignment / bytesPerBlock;
+
+    const expectedData = new ArrayBuffer(
+      bytesPerRowAlignment * (height * arrayLayers - 1) + bytesPerBlock * width
+    );
+    for (let z = 0; z < arrayLayers; ++z) {
+      for (let y = 0; y < height; ++y) {
+        for (let x = 0; x < width; ++x) {
+          const expectedIndex = z * itemsPerRow * height + y * itemsPerRow + x;
+          const initialIndex =
+            (arrayLayers - 1 - z) * width * height + (height - 1 - y) * width + (width - 1 - x);
+          switch (textureFormat) {
+            case 'r32sint': {
+              const expectedSintData = new Int32Array(expectedData);
```

Comment:
This might create a LOT of garbage, maybe we can declare them outside of the loop? GC pauses have been a source of flakiness and slowness in tests sometimes.